[Naaman Tan] iP

tanyjnaaman opened the pull request:
# Duke's successor, Alfred

## What is the purpose of this pull request?

Well, get a good grade in CS2103. Also, improve Duke. 

## What is the rationale for this pull request?

- Implement core features of Alfred, including adding task, mark as done and delete functionalities, for _ToDo_, _Event_ and _Deadline_ task types.
- Build core functionalities of Alfred with proper use of OOP and unit testing
- Implement save functionalities of Alfred
- Implement exception handling for erroneous user input 
- Build a GUI 

What changes did I make?

Encapsulated core logic of adding, marking, deleting tasks into respective `Task` and `Command` classes, save/load functionalities into `Storage` class, user-related interactions into `UI` and exceptions into `Exceptions` and used a package file structure for neatness. 

At a high level, Alfred's logic is as follows:

```java

    public void run() {
        this.ui.greeting();
        boolean isExit = false;
        while (!isExit) {
            try {
                String input = ui.readInput();
                Command command = this.parser.parseInput(input);
                command.execute(ui, storage);
                isExit = command.isExit();
            } catch (AlfredException e) {
                ui.handleAlfredException(e);
            }
        }
    }

```

## What changes are left to be made?
At the end, the aim is to create a functioning, *simple* chatbot with save functionalities and a UI, that is *built well*. 🥇 

The design philosophy follows this quote:

> “Your mind is for having ideas, not holding them.” – David Allen ( [source](https://dansilvestre.com/productivity-quotes) )

It's cringey, but I had to use block quotes somewhere.

1. Automate project build using Gradle
2. UI (using JavaFX)

## Summary of progress

- [X] Mark as done, delete functionalities
- [X] Task types
- [X] Exception handling
- [X] Save functionality
- [X] Refactor for OOP
- [x] Gradle
- [x] Window UI

tanyjnaaman committed:
Renamed to Alfred. Implemented level one.

tanyjnaaman committed:
Add, List functionalities added.

tanyjnaaman committed:
Changed privacy of methods

tanyjnaaman committed:
Added mark as done functionality

tanyjnaaman committed:
Added ToDos, Events and Deadlines functionality

tanyjnaaman committed:
Adding test cases up to level 4

tanyjnaaman committed:
Included exception handling for user inputs

tanyjnaaman committed:
Updated test cases

tanyjnaaman committed:
Added delete functionality.

tanyjnaaman committed:
Fixed deadline and event to prevent empty tasks

tanyjnaaman committed:
Fix minor bugs

tanyjnaaman committed:
Implement save to .txt feature.

* Refactor variable and method names to follow coding style
* Split Alfred::redaInput method into readInput, which handles
  sanitisation and exception handling, and callCommand, which handles
  calling methods and saving to disk.

tanyjnaaman committed:
Refactor code for style and store dates as LocalDate

tanyjnaaman committed:
Merge branch 'branch-Level-8'

tanyjnaaman committed:
Refactor the code to extract closely related code as classes.

Organised into:
* Command
* Exception
* Parser
* Storage
* Alfred (itself)

And execution is in the Run class.

tanyjnaaman committed:
Add JUnit tests to test the behaviour of the code.

Created two tests, one for
Alfred.UI.AlfredUserInterface::sandwichAndPrint, and one for
Alfred.Task.ToDo:toString

tanyjnaaman committed:
Add JavaDoc comments to the code.

tanyjnaaman committed:
Modified code to comply with coding standard

tanyjnaaman committed:
Implement keyword search

tanyjnaaman committed:
Implement keyword search - previous did not stage

tanyjnaaman committed:
Merge branch 'branch-Level-9'

tanyjnaaman committed:
Merge branch 'branch-A-javadoc'

tanyjnaaman committed:
Merge branch 'branch-A-CodingStandard'

tanyjnaaman committed:
Tweak code to comply with coding standard

tanyjnaaman committed:
Merge branch 'branch-undo-delete'

tanyjnaaman committed:
rename subfolders to fit naming convention

tanyjnaaman committed:
Fix Level-7 save bug

tanyjnaaman committed:
Automate project builds and unit test using Gradle

tanyjnaaman committed:
Include checkstyle in build.gradle

tanyjnaaman committed:
add a GUI to Alfred

tanyjnaaman committed:
Update README.md

onefangg commented on src/main/java/alfred/command/DeadlineCommand.java:
```diff
@@ -0,0 +1,78 @@
+    }
+
+    @Override
+    public String response(AlfredUserInterface ui, AlfredStorage storage)
+            throws InvalidDateTimeException, InvalidInputException {
```
Missing header comments for overwritten methods from subclasses of Command - maybe you can try `@inheritDoc` as mentioned in the form post [here](https://github.com/nus-cs2103-AY2122S2/forum/issues/71)

onefangg commented on src/main/java/alfred/parser/AlfredParser.java:
```diff
@@ -0,0 +1,66 @@
+        if ((command.equals("list")) && (arguments.length == 1)) {
+            return new ListCommand();
+
+            // (UN)MARK and DELETE
```
Extra indentation at comment

onefangg commented on src/main/java/alfred/ui/controller/DialogBox.java:
```diff
@@ -0,0 +1,74 @@
+        setAlignment(Pos.TOP_LEFT);
+    }
+
+    public static DialogBox getUserDialog(String text, Image img) {
+        var db = new DialogBox(text, img);
+        db.flip();
+        return db;
+    }
+
+    public static DialogBox getAlfredDialog(String text, Image img) {
+        return new DialogBox(text, img);
+    }
```
Missing header comments

onefangg commented on src/main/java/alfred/ui/controller/MainWindow.java:
```diff
@@ -0,0 +1,65 @@
+        this.scrollPane.vvalueProperty().bind(this.dialogContainer.heightProperty());
+    }
+
+    public void greetUser() {
```
Missing header comments

onefangg commented on src/main/java/alfred/task/Task.java:
```diff
@@ -0,0 +1,119 @@
+
+    // instance attributes
+    protected final String description;
+    protected boolean completed;
```
Maybe include a prefix to name so to make it sound like boolean?

yumengtan replied:
Perhaps using a prefix such as is to the variable name such as IsCompleted would make it sound more like booleans!

onefangg reviewed (commented):
Looks good - just minor cosmetic changes needed

yumengtan commented on src/main/java/alfred/storage/AlfredStorage.java:
```diff
@@ -0,0 +1,227 @@
+        }
+    }
+
+    public String summarizeList() {
```
Perhaps adding header comments would be easier to understand what the method does at a glance!

yumengtan reviewed (commented):
Overall your code structure is very well organised! Love the twist to Duke u have added! Just some nits to fix, LGTM!

tanyjnaaman committed:
Refactor to improve coding standard

tanyjnaaman committed:
Merge branch 'master' of https://github.com/tanyjnaaman/ip

tanyjnaaman committed:
Use assertions to document important assumptions

tanyjnaaman committed:
Merge pull request #1 from tanyjnaaman/branch-A-Assertions

Use assertions to document important assumptions

tanyjnaaman committed:
Improve code quality

Refactored methods to improve code quality. In particular, focused on
applying SLAP as far as possible, using guard clauses, reducing
arrow-shaped blocks, removing magic literals and writing comments
according to WHAT and WHY, removing HOW.

tanyjnaaman committed:
Merge pull request #2 from tanyjnaaman/branch-A-CodeQuality

Improve code quality

tanyjnaaman committed:
Set up continuous integration

Set up CI with gradle.yml file. Also fixed relevant formatting issues
flagged by gradle and added a checkstyle.xml file.

tanyjnaaman committed:
Merge pull request #3 from tanyjnaaman/branch-A-CI

Set up continuous integration

tanyjnaaman committed:
Add functionality to deal with duplicates

Add the ability to recognize and handle duplicate items by providing
feedback to user.

tanyjnaaman committed:
Merge pull request #4 from tanyjnaaman/branch-C-DetectDuplicates

Add functionality to deal with duplicates

tanyjnaaman committed:
Add functionality to deal with duplicates

Add the ability to recognize and handle duplicate items by providing
feedback to the user. Previous commit did not have staged changes.

tanyjnaaman committed:
Merge pull request #5 from tanyjnaaman/branch-C-DetectDuplicates

Add functionality to deal with duplicates

tanyjnaaman committed:
improve the GUI

Tweaked the GUI to make it more polished. Previously, it had
fixed-size text boxes independent of length of text, and the
representation of user and Alfred text boxes were identical,
making it hard to read.

In response, changes are:
* include different colour text boxes for user and alfred
* dynamically size text boxes
* use of .css file for future extendibility
* rounding of user pictures

tanyjnaaman committed:
improve GUI further

GUI felt bland and hard to read. Previous changes tackled concrete
things like dynamic text resizing, and the like. This commit focuses
more on style. In particular, the changes made are:

* Implement "Alfred" header
* Modify representation of task types, using emojis for better
  visualization

* Use of contrasting colours (white on blue) for Alfred's replies, so
  that they are easier to read

tanyjnaaman committed:
add user guide

As the features are finalize, a user guide is prepared. Here, a brief
draft of the user guide is made.

tanyjnaaman committed:
add a user guide

Add a user guide to the project.

tanyjnaaman committed:
format docs/README.md for easier reading
